File: MemSharp/clients/go/client_test.go
package memsharp_test

// Integration tests for the Go client.
//
// These run against a live server rather than a mock, because the thing worth testing is that the
// wire encoding matches what MemSharp actually sends back. Start one first:
//
//	memsharp serve --port 6391 --quiet
//	go test ./clients/go/...
//
// Without a server the tests skip rather than fail, so `go test ./...` on a fresh checkout stays
// green.

import (
	"os"
	"sort"
	"testing"
	"time"

	memsharp "github.com/DotNetVibeCoderz/Vibe_Database/MemSharp/clients/go"
)

func address() string {
	if value := os.Getenv("MEMSHARP_TEST_ADDRESS"); value != "" {
		return value
	}
	return "127.0.0.1:6391"
}

// connect dials the test server, skipping the test when none is running.
func connect(t *testing.T) *memsharp.Client {
	t.Helper()

	db, err := memsharp.Dial(address())
	if err != nil {
		t.Skipf("no MemSharp server at %s: %v", address(), err)
	}
	t.Cleanup(func() { _ = db.Close() })
	return db
}

func TestStrings(t *testing.T) {
	db := connect(t)

	if err := db.Set("go:k", "v"); err != nil {
		t.Fatalf("Set: %v", err)
	}

	value, found, err := db.Get("go:k")
	if err != nil || !found || value != "v" {
		t.Fatalf("Get = %q, %v, %v; want \"v\", true, nil", value, found, err)
	}

	if _, found, _ := db.Get("go:absent"); found {
		t.Error("Get on a missing key reported found")
	}

	if n, err := db.Incr("go:n", 41); err != nil || n != 41 {
		t.Fatalf("Incr = %d, %v; want 41", n, err)
	}
	if n, err := db.Incr("go:n", 1); err != nil || n != 42 {
		t.Fatalf("Incr = %d, %v; want 42", n, err)
	}

	if f, err := db.IncrByFloat("go:f", 1.5); err != nil || f != 1.5 {
		t.Fatalf("IncrByFloat = %v, %v; want 1.5", f, err)
	}

	_, _ = db.Del("go:k", "go:n", "go:f")
}

func TestMGetPreservesMissingKeys(t *testing.T) {
	db := connect(t)

	_ = db.Set("go:a", "1")
	_ = db.Set("go:c", "3")
	defer db.Del("go:a", "go:c")

	values, err := db.MGet("go:a", "go:b", "go:c")
	if err != nil {
		t.Fatalf("MGet: %v", err)
	}
	if len(values) != 3 {
		t.Fatalf("MGet returned %d values; want 3", len(values))
	}

	// A nil pointer rather than an empty string, so a missing key stays distinguishable from a key
	// holding "".
	if values[0] == nil || *values[0] != "1" {
		t.Errorf("values[0] = %v; want \"1\"", values[0])
	}
	if values[1] != nil {
		t.Errorf("values[1] = %v; want nil for the missing key", *values[1])
	}
	if values[2] == nil || *values[2] != "3" {
		t.Errorf("values[2] = %v; want \"3\"", values[2])
	}
}

func TestExpiry(t *testing.T) {
	db := connect(t)

	if err := db.SetEx("go:temp", "v", time.Minute); err != nil {
		t.Fatalf("SetEx: %v", err)
	}

	ttl, hasTTL, err := db.TTL("go:temp")
	if err != nil || !hasTTL || ttl > time.Minute || ttl < 55*time.Second {
		t.Fatalf("TTL = %v, %v, %v; want about a minute", ttl, hasTTL, err)
	}

	_ = db.Set("go:permanent", "v")
	if _, hasTTL, _ := db.TTL("go:permanent"); hasTTL {
		t.Error("a permanent key reported a TTL")
	}

	if cleared, err := db.Persist("go:temp"); err != nil || !cleared {
		t.Fatalf("Persist = %v, %v; want true", cleared, err)
	}

	_ = db.SetEx("go:brief", "v", 50*time.Millisecond)
	time.Sleep(200 * time.Millisecond)
	if _, found, _ := db.Get("go:brief"); found {
		t.Error("an expired key was still readable")
	}

	_, _ = db.Del("go:temp", "go:permanent")
}

func TestSetNX(t *testing.T) {
	db := connect(t)
	defer db.Del("go:nx")

	if stored, err := db.SetNX("go:nx", "first"); err != nil || !stored {
		t.Fatalf("SetNX on an absent key = %v, %v; want true", stored, err)
	}
	if stored, err := db.SetNX("go:nx", "second"); err != nil || stored {
		t.Fatalf("SetNX on an existing key = %v, %v; want false", stored, err)
	}

	value, _, _ := db.Get("go:nx")
	if value != "first" {
		t.Errorf("value = %q; want \"first\"", value)
	}
}

func TestLists(t *testing.T) {
	db := connect(t)
	defer db.Del("go:list")

	if n, err := db.RPush("go:list", "a", "b", "c"); err != nil || n != 3 {
		t.Fatalf("RPush = %d, %v; want 3", n, err)
	}

	items, err := db.LRange("go:list", 0, -1)
	if err != nil || len(items) != 3 || items[0] != "a" || items[2] != "c" {
		t.Fatalf("LRange = %v, %v; want [a b c]", items, err)
	}

	// Negative indices count back from the end.
	tail, err := db.LRange("go:list", -2, -1)
	if err != nil || len(tail) != 2 || tail[0] != "b" {
		t.Fatalf("LRange(-2, -1) = %v, %v; want [b c]", tail, err)
	}

	if _, _ = db.LPush("go:list", "z"); true {
		if head, found, _ := db.LPop("go:list"); !found || head != "z" {
			t.Errorf("LPop = %q, %v; want \"z\", true", head, found)
		}
	}

	if last, found, _ := db.RPop("go:list"); !found || last != "c" {
		t.Errorf("RPop = %q, %v; want \"c\", true", last, found)
	}

	if err := db.LTrim("go:list", 0, 0); err != nil {
		t.Fatalf("LTrim: %v", err)
	}
	if n, _ := db.LLen("go:list"); n != 1 {
		t.Errorf("LLen after LTrim = %d; want 1", n)
	}
}

func TestHashes(t *testing.T) {
	db := connect(t)
	defer db.Del("go:hash")

	if _, err := db.HSet("go:hash", map[string]string{"name": "Kang Fadhil", "desk": "Jakarta"}); err != nil {
		t.Fatalf("HSet: %v", err)
	}

	fields, err := db.HGetAll("go:hash")
	if err != nil || fields["name"] != "Kang Fadhil" || fields["desk"] != "Jakarta" {
		t.Fatalf("HGetAll = %v, %v", fields, err)
	}

	if value, found, _ := db.HGet("go:hash", "desk"); !found || value != "Jakarta" {
		t.Errorf("HGet = %q, %v; want \"Jakarta\", true", value, found)
	}
	if _, found, _ := db.HGet("go:hash", "absent"); found {
		t.Error("HGet on a missing field reported found")
	}

	if n, err := db.HIncrBy("go:hash", "fills", 5); err != nil || n != 5 {
		t.Fatalf("HIncrBy = %d, %v; want 5", n, err)
	}
	if n, err := db.HDel("go:hash", "desk"); err != nil || n != 1 {
		t.Fatalf("HDel = %d, %v; want 1", n, err)
	}
}

func TestSets(t *testing.T) {
	db := connect(t)
	defer db.Del("go:set")

	if n, err := db.SAdd("go:set", "x", "y", "x"); err != nil || n != 2 {
		t.Fatalf("SAdd = %d, %v; want 2 - duplicates should not count", n, err)
	}

	members, err := db.SMembers("go:set")
	sort.Strings(members)
	if err != nil || len(members) != 2 || members[0] != "x" || members[1] != "y" {
		t.Fatalf("SMembers = %v, %v; want [x y]", members, err)
	}

	if has, _ := db.SIsMember("go:set", "x"); !has {
		t.Error("SIsMember said x is absent")
	}
	if has, _ := db.SIsMember("go:set", "q"); has {
		t.Error("SIsMember said q is present")
	}
	if n, _ := db.SRem("go:set", "x"); n != 1 {
		t.Errorf("SRem = %d; want 1", n)
	}
}

func TestSortedSets(t *testing.T) {
	db := connect(t)
	defer db.Del("go:zset")

	_, err := db.ZAdd("go:zset",
		memsharp.ScoredMember{Member: "low", Score: 1},
		memsharp.ScoredMember{Member: "high", Score: 3},
		memsharp.ScoredMember{Member: "mid", Score: 2})
	if err != nil {
		t.Fatalf("ZAdd: %v", err)
	}

	ascending, err := db.ZRange("go:zset", 0, -1, false)
	if err != nil || len(ascending) != 3 || ascending[0] != "low" || ascending[2] != "high" {
		t.Fatalf("ZRange = %v, %v; want [low mid high]", ascending, err)
	}

	descending, _ := db.ZRange("go:zset", 0, -1, true)
	if len(descending) != 3 || descending[0] != "high" {
		t.Fatalf("ZRange descending = %v; want [high mid low]", descending)
	}

	scored, err := db.ZRangeWithScores("go:zset", 0, 0, false)
	if err != nil || len(scored) != 1 || scored[0].Member != "low" || scored[0].Score != 1 {
		t.Fatalf("ZRangeWithScores = %v, %v", scored, err)
	}

	window, _ := db.ZRangeByScore("go:zset", 1.5, 2.5)
	if len(window) != 1 || window[0] != "mid" {
		t.Fatalf("ZRangeByScore = %v; want [mid]", window)
	}

	if score, found, _ := db.ZScore("go:zset", "high"); !found || score != 3 {
		t.Errorf("ZScore = %v, %v; want 3, true", score, found)
	}
	if _, found, _ := db.ZScore("go:zset", "absent"); found {
		t.Error("ZScore on a missing member reported found")
	}
	if n, _ := db.ZCard("go:zset"); n != 3 {
		t.Errorf("ZCard = %d; want 3", n)
	}
}

func TestStreams(t *testing.T) {
	db := connect(t)
	defer db.Del("go:stream", "go:capped")

	id, err := db.XAdd("go:stream", map[string]string{"sym": "BTC", "qty": "5"}, 0)
	if err != nil || id == "" {
		t.Fatalf("XAdd = %q, %v", id, err)
	}
	_, _ = db.XAdd("go:stream", map[string]string{"sym": "ETH", "qty": "12"}, 0)

	if n, _ := db.XLen("go:stream"); n != 2 {
		t.Fatalf("XLen = %d; want 2", n)
	}

	entries, err := db.XRange("go:stream", "-", "+", 0)
	if err != nil || len(entries) != 2 {
		t.Fatalf("XRange = %v, %v; want 2 entries", entries, err)
	}
	if entries[0].Fields["sym"] != "BTC" || entries[0].Fields["qty"] != "5" {
		t.Errorf("entries[0].Fields = %v", entries[0].Fields)
	}

	for i := 0; i < 50; i++ {
		_, _ = db.XAdd("go:capped", map[string]string{"n": "x"}, 10)
	}
	if n, _ := db.XLen("go:capped"); n != 10 {
		t.Errorf("XLen with MAXLEN 10 = %d; want 10", n)
	}
}

func TestTimeSeries(t *testing.T) {
	db := connect(t)

	// A fresh key each run: a series left over from a previous run holds a server-stamped sample,
	// and the small timestamps below would then be rejected as out of order - correctly.
	key := "go:ts"
	_, _ = db.Del(key)
	defer db.Del(key)

	if err := db.TSCreate(key, 1000); err != nil {
		t.Fatalf("TSCreate: %v", err)
	}

	for i := int64(0); i < 20; i++ {
		if _, err := db.TSAdd(key, float64(i), i*100); err != nil {
			t.Fatalf("TSAdd: %v", err)
		}
	}

	// TSAddNow uses the server clock, so it lands after every explicit timestamp above.
	if stamped, err := db.TSAddNow(key, 99); err != nil || stamped <= 1900 {
		t.Fatalf("TSAddNow = %d, %v; want a server timestamp", stamped, err)
	}

	samples, err := db.TSRange(key, 0, 10_000)
	if err != nil || len(samples) != 20 {
		t.Fatalf("TSRange = %d samples, %v; want 20", len(samples), err)
	}
	if samples[0].Timestamp != 0 || samples[0].Value != 0 {
		t.Errorf("samples[0] = %+v; want {0 0}", samples[0])
	}

	buckets, err := db.TSAggregate(key, 0, 10_000, 500, "max")
	if err != nil || len(buckets) != 4 {
		t.Fatalf("TSAggregate = %v, %v; want 4 buckets", buckets, err)
	}
	if buckets[0].Value != 4 {
		t.Errorf("buckets[0].Value = %v; want 4", buckets[0].Value)
	}
}

func TestSQL(t *testing.T) {
	db := connect(t)

	for i := 0; i < 20; i++ {
		_ = db.Set("go:order:"+string(rune('a'+i)), "value")
	}
	defer db.SQLDelete("DELETE FROM keys WHERE key LIKE 'go:order:%'")

	rows, err := db.SQL("SELECT key, type FROM keys WHERE key LIKE 'go:order:%' ORDER BY key LIMIT 5")
	if err != nil {
		t.Fatalf("SQL: %v", err)
	}
	if len(rows) != 5 {
		t.Fatalf("SQL returned %d rows; want 5", len(rows))
	}

	// Rows come back keyed by column name rather than by position.
	if rows[0]["type"] != "String" || rows[0]["key"] == "" {
		t.Errorf("rows[0] = %v; want key and type columns", rows[0])
	}

	removed, err := db.SQLDelete("DELETE FROM keys WHERE key LIKE 'go:order:%'")
	if err != nil || removed != 20 {
		t.Fatalf("SQLDelete = %d, %v; want 20", removed, err)
	}
}

func TestErrorsCarryTheirCode(t *testing.T) {
	db := connect(t)
	defer db.Del("go:errlist")

	_, _ = db.RPush("go:errlist", "x")

	_, _, err := db.Get("go:errlist")
	if err == nil {
		t.Fatal("Get on a list did not fail")
	}
	if !memsharp.IsWrongType(err) {
		t.Errorf("IsWrongType(%v) = false; want true", err)
	}

	// An error must not desynchronise the connection.
	if reply, err := db.Ping(); err != nil || reply != "PONG" {
		t.Errorf("Ping after an error = %q, %v; want PONG", reply, err)
	}
}

func TestPipeline(t *testing.T) {
	db := connect(t)

	commands := make([][]any, 0, 500)
	for i := 0; i < 500; i++ {
		commands = append(commands, []any{"SET", "go:p" + string(rune(i%26+'a')) + string(rune(i/26+'a')), i})
	}

	replies, err := db.Pipeline(commands)
	if err != nil {
		t.Fatalf("Pipeline: %v", err)
	}
	if len(replies) != len(commands) {
		t.Fatalf("Pipeline returned %d replies; want %d", len(replies), len(commands))
	}
	for i, reply := range replies {
		if reply != "OK" {
			t.Fatalf("replies[%d] = %v; want OK", i, reply)
		}
	}

	// Errors come back in place rather than aborting the batch.
	mixed, err := db.Pipeline([][]any{{"PING"}, {"NOSUCHCOMMAND"}, {"PING"}})
	if err != nil {
		t.Fatalf("Pipeline with a bad command: %v", err)
	}
	if mixed[0] != "PONG" || mixed[2] != "PONG" {
		t.Errorf("surrounding replies = %v, %v; want PONG", mixed[0], mixed[2])
	}
	if _, isError := mixed[1].(*memsharp.Error); !isError {
		t.Errorf("mixed[1] = %T; want *memsharp.Error", mixed[1])
	}

	_, _ = db.SQLDelete("DELETE FROM keys WHERE key LIKE 'go:p%'")
}

func TestScanVisitsEveryKey(t *testing.T) {
	db := connect(t)

	for i := 0; i < 200; i++ {
		_ = db.Set("go:scan:"+string(rune('a'+i%26))+string(rune('a'+i/26)), "v")
	}
	defer db.SQLDelete("DELETE FROM keys WHERE key LIKE 'go:scan:%'")

	seen := make(map[string]bool)
	err := db.Scan("go:scan:*", 32, func(key string) error {
		seen[key] = true
		return nil
	})
	if err != nil {
		t.Fatalf("Scan: %v", err)
	}
	if len(seen) != 200 {
		t.Errorf("Scan visited %d distinct keys; want 200", len(seen))
	}
}

func TestPubSub(t *testing.T) {
	db := connect(t)

	subscriber, err := memsharp.Dial(address())
	if err != nil {
		t.Skipf("no server: %v", err)
	}
	defer subscriber.Close()

	received := make(chan string, 4)
	done := make(chan struct{})

	go func() {
		defer close(done)
		_ = subscriber.Subscribe(func(message memsharp.Message) error {
			// The warmup publishes below only detect that the subscription registered.
			if message.Payload == "warmup" {
				return nil
			}
			received <- message.Payload
			if len(received) == 2 {
				return errStopSubscription
			}
			return nil
		}, "go:news")
	}()

	deadline := time.Now().Add(5 * time.Second)
	for time.Now().Before(deadline) {
		if n, _ := db.Publish("go:news", "warmup"); n > 0 {
			break
		}
		time.Sleep(50 * time.Millisecond)
	}

	_, _ = db.Publish("go:news", "one")
	_, _ = db.Publish("go:news", "two")

	select {
	case <-done:
	case <-time.After(5 * time.Second):
		t.Fatal("pub/sub messages did not arrive")
	}

	close(received)
	var payloads []string
	for payload := range received {
		payloads = append(payloads, payload)
	}
	if len(payloads) != 2 || payloads[0] != "one" || payloads[1] != "two" {
		t.Errorf("payloads = %v; want [one two]", payloads)
	}
}

// errStopSubscription ends a Subscribe loop from inside the callback.
var errStopSubscription = &memsharp.Error{Code: "STOP", Message: "test finished"}

func TestServerInfo(t *testing.T) {
	db := connect(t)

	info, err := db.Info()
	if err != nil {
		t.Fatalf("Info: %v", err)
	}
	if info["product"] != "MemSharp" {
		t.Errorf("info[product] = %q; want MemSharp", info["product"])
	}
	if _, present := info["keys"]; !present {
		t.Error("info has no keys entry")
	}

	if _, err := db.DBSize(); err != nil {
		t.Errorf("DBSize: %v", err)
	}
}

File: MemSharp/clients/nodejs/test/client.test.js
'use strict';

/**
 * Integration tests for the Node.js client.
 *
 * These run against a live server rather than a mock, because the thing worth testing is that the
 * wire encoding matches what MemSharp actually sends back. Start one first:
 *
 *   memsharp serve --port 6391 --quiet
 *   node clients/nodejs/test/client.test.js
 */

const { MemSharpClient, WrongTypeError } = require('../index');

const PORT = Number(process.env.MEMSHARP_TEST_PORT || 6391);
const failures = [];

function check(name, condition, detail = '') {
  const status = condition ? '  ok  ' : ' FAIL ';
  console.log(`${status} ${name}${detail ? `  [${detail}]` : ''}`);
  if (!condition) failures.push(name);
}

function same(a, b) {
  return JSON.stringify(a) === JSON.stringify(b);
}

async function main() {
  const db = new MemSharpClient({ port: PORT });
  await db.connect();
  await db.flushdb();

  // -- strings and counters --------------------------------------------------------
  check('ping', (await db.ping()) === 'PONG');
  await db.set('k', 'v');
  check('set/get', (await db.get('k')) === 'v');
  check('get missing', (await db.get('absent')) === null);
  check('incr', (await db.incr('n', 41)) === 41 && (await db.incr('n')) === 42);
  check('incrByFloat', Math.abs((await db.incrByFloat('f', 1.5)) - 1.5) < 1e-9);
  check('mget order', same(await db.mget('k', 'absent', 'n'), ['v', null, '42']));

  await db.mset({ a: '1', b: '2' });
  check('mset', same(await db.mget('a', 'b'), ['1', '2']));
  check('exists', (await db.exists('a', 'b', 'nope')) === 2);
  check('type', (await db.type('k')) === 'string');
  check('del', (await db.del('a', 'b')) === 2);

  // -- expiry ----------------------------------------------------------------------
  await db.set('temp', 'v', { ex: 60 });
  const ttl = await db.ttl('temp');
  check('ttl set', ttl !== null && ttl > 55 && ttl <= 60, String(ttl));
  check('ttl permanent', (await db.ttl('k')) === null);
  check('persist', (await db.persist('temp')) && (await db.ttl('temp')) === null);

  await db.set('brief', 'v', { ex: 0.05 });
  await new Promise((resolve) => setTimeout(resolve, 200));
  check('expiry evicts', (await db.get('brief')) === null);

  check('set nx', (await db.set('k', 'other', { nx: true })) === false && (await db.get('k')) === 'v');

  // -- lists -----------------------------------------------------------------------
  await db.rpush('l', 'a', 'b', 'c');
  check('rpush/lrange', same(await db.lrange('l'), ['a', 'b', 'c']));
  await db.lpush('l', 'z');
  check('lpush', same(await db.lrange('l', 0, 0), ['z']));
  check('llen', (await db.llen('l')) === 4);
  check('lpop', (await db.lpop('l')) === 'z');
  check('rpop', (await db.rpop('l')) === 'c');
  await db.ltrim('l', 0, 0);
  check('ltrim', same(await db.lrange('l'), ['a']));

  await db.rpush('r', '1', '2', '3');
  check('negative index range', same(await db.lrange('r', -2, -1), ['2', '3']));

  // -- hashes ----------------------------------------------------------------------
  await db.hset('h', { name: 'Kang Fadhil', desk: 'Jakarta' });
  check('hset/hgetall', same(await db.hgetall('h'), { name: 'Kang Fadhil', desk: 'Jakarta' }));
  check('hget', (await db.hget('h', 'desk')) === 'Jakarta');
  check('hincrby', (await db.hincrby('h', 'fills', 5)) === 5);
  check('hdel', (await db.hdel('h', 'desk')) === 1 && (await db.hget('h', 'desk')) === null);

  // -- sets ------------------------------------------------------------------------
  await db.sadd('s', 'x', 'y', 'x');
  check('sadd dedupes', (await db.smembers('s')).sort().join(',') === 'x,y');
  check('sismember', (await db.sismember('s', 'x')) && !(await db.sismember('s', 'q')));
  check('srem', (await db.srem('s', 'x')) === 1);

  // -- sorted sets -----------------------------------------------------------------
  await db.zadd('z', { low: 1.0, high: 3.0, mid: 2.0 });
  check('zadd/zrange', same(await db.zrange('z'), ['low', 'mid', 'high']));
  check('zrange desc', same(await db.zrange('z', 0, -1, { desc: true }), ['high', 'mid', 'low']));

  const scored = await db.zrange('z', 0, -1, { withScores: true });
  check('zrange withScores', same(scored[0], { member: 'low', score: 1 }), JSON.stringify(scored[0]));
  check('zrangeByScore', same(await db.zrangeByScore('z', 1.5, 2.5), ['mid']));
  check('zscore', (await db.zscore('z', 'high')) === 3);
  check('zscore missing', (await db.zscore('z', 'nope')) === null);
  check('zcard', (await db.zcard('z')) === 3);

  // -- streams ---------------------------------------------------------------------
  const first = await db.xadd('stream', { sym: 'BTC', qty: '5' });
  await db.xadd('stream', { sym: 'ETH', qty: '12' });
  check('xadd returns id', typeof first === 'string' && first.includes('-'), first);
  check('xlen', (await db.xlen('stream')) === 2);

  const entries = await db.xrange('stream');
  check('xrange fields', same(entries[0].fields, { sym: 'BTC', qty: '5' }), JSON.stringify(entries[0].fields));

  for (let i = 0; i < 50; i++) await db.xadd('capped', { n: String(i) }, { maxLen: 10 });
  check('xadd maxLen', (await db.xlen('capped')) === 10);

  // -- time series -----------------------------------------------------------------
  await db.tsCreate('ts', 1000);
  for (let i = 0; i < 20; i++) await db.tsAdd('ts', i, i * 100);

  const samples = await db.tsRange('ts', 0, 10000);
  check('tsRange', samples.length === 20 && samples[0].timestamp === 0, JSON.stringify(samples[0]));

  const buckets = await db.tsAggregate('ts', 0, 10000, 500, 'max');
  check('tsAggregate', buckets.length === 4 && buckets[0].value === 4, JSON.stringify(buckets));

  // -- SQL -------------------------------------------------------------------------
  for (let i = 0; i < 20; i++) await db.set(`order:${i}`, 'x'.repeat(i));

  const rows = await db.sql("SELECT key, size FROM keys WHERE key LIKE 'order:%' AND size > 15 ORDER BY size DESC LIMIT 3");
  check('sql rows are objects', rows.length === 3 && rows[0].size === '19', JSON.stringify(rows[0]));
  check('sql column names', same(Object.keys(rows[0]).sort(), ['key', 'size']));
  check('sql delete', (await db.sqlDelete("DELETE FROM keys WHERE key LIKE 'order:1%'")) === 11);

  // -- errors ----------------------------------------------------------------------
  try {
    await db.get('l');
    check('wrongtype rejects', false);
  } catch (error) {
    check('wrongtype rejects', error instanceof WrongTypeError && error.code === 'WRONGTYPE', error.detail);
  }

  // An error must not desynchronise the connection.
  check('connection survives an error', (await db.ping()) === 'PONG');

  // -- pipelining ------------------------------------------------------------------
  const batch = [];
  for (let i = 0; i < 500; i++) batch.push(['SET', `p${i}`, String(i)]);

  const replies = await db.pipeline(batch);
  check('pipeline replies', replies.length === 500 && replies.every((r) => r === 'OK'));
  check('pipeline stored', (await db.get('p499')) === '499');

  const mixed = await db.pipeline([['PING'], ['NOSUCHCOMMAND'], ['PING']]);
  check('pipeline returns errors in place',
    mixed[0] === 'PONG' && mixed[1] instanceof Error && mixed[2] === 'PONG');

  // -- scan ------------------------------------------------------------------------
  const found = [];
  for await (const key of db.scan('p*', 64)) found.push(key);
  check('scan pages', found.length === 500, String(found.length));

  // -- pub/sub ---------------------------------------------------------------------
  const subscriber = new MemSharpClient({ port: PORT });
  await subscriber.connect();

  const received = [];
  const delivered = new Promise((resolve) => {
    subscriber.on('message', (channel, message) => {
      // The warmup publishes below only detect that the subscription registered.
      if (message === 'warmup') return;
      received.push(message);
      if (received.length === 2) resolve();
    });
  });

  await subscriber.subscribe('news');

  const deadline = Date.now() + 5000;
  while ((await db.publish('news', 'warmup')) === 0 && Date.now() < deadline) {
    await new Promise((resolve) => setTimeout(resolve, 50));
  }

  await db.publish('news', 'one');
  await db.publish('news', 'two');
  await Promise.race([delivered, new Promise((resolve) => setTimeout(resolve, 5000))]);

  check('pubsub delivers', same(received, ['one', 'two']), JSON.stringify(received));
  await subscriber.close();

  // -- server ----------------------------------------------------------------------
  const info = await db.info();
  check('info parses', info.product === 'MemSharp' && 'keys' in info);
  check('dbsize', (await db.dbsize()) > 0);

  await db.close();

  console.log();
  console.log(failures.length === 0 ? 'ALL PASS' : `${failures.length} FAILURE(S): ${failures.join(', ')}`);
  process.exit(failures.length === 0 ? 0 : 1);
}

main().catch((error) => {
  console.error('test run failed:', error);
  process.exit(1);
});
